include/pybind11/functional.h

```diff
@@ -38,19 +38,30 @@ struct type_caster<std::function<Return(Args...)>> {
 
         auto func = reinterpret_borrow<function>(src);
 
-        /*
-           When passing a C++ function as an argument to another C++
-           function via Python, every function call would normally involve
-           a full C++ -> Python -> C++ roundtrip, which can be prohibitive.
-           Here, we try to at least detect the case where the function is
-           stateless (i.e. function pointer or lambda function without
-           captured variables), in which case the roundtrip can be avoided.
-         */
+/*
+   When passing a C++ function as an argument to another C++
+   function via Python, every function call would normally involve
+   a full C++ -> Python -> C++ roundtrip, which can be prohibitive.
+   Here, we try to at least detect the case where the function is
+   stateless (i.e. function pointer or lambda function without
+   captured variables), in which case the roundtrip can be avoided.
+ */
+#ifndef BAD
         if (auto cfunc = func.cpp_function()) {
             auto *cfunc_self = PyCFunction_GET_SELF(cfunc.ptr());
-            if (isinstance<capsule>(cfunc_self)) {
+            if (cfunc_self == nullptr) {
+                PyErr_Clear();
+            } else if (isinstance<capsule>(cfunc_self)) {
                 auto c = reinterpret_borrow<capsule>(cfunc_self);
-                auto *rec = (function_record *) c;
+
+                function_record *rec = nullptr;
+                // THIS IS THE PROBLEM: UNSAFE NO WAY TO FIGURE OUT IF CAPSULE IS FOREIGN
```

Comment:
This is actually the problem in several places in pybind11.h, but hitting a custom def is much rarer there.